pkg/rhos/ocpgwdeployer.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		return errors.WithMessage(err, "error creating the network client")
+	}
+
+	groupName := d.InfraID + gwSecurityGroupSuffix
+	if err := d.createGWSecurityGroup(input.PublicPorts, groupName, computeClient, networkClient); err != nil {
+		return errors.WithMessage(err, "creating gateway security group failed")
+	}
+
+	reporter.Succeeded("Opened External ports %q in security group %q on RHOS",
+		formatPorts(input.PublicPorts), groupName)
+
+	reporter.Started("Configuring the required number of gateway pods")
+
+	gwNodes, err := d.K8sClient.ListGatewayNodes()
+	if err != nil {
+		return errors.WithMessagef(err, "Listing the existing gatway nodes failed")
+	}
+
+	numGatewayNodes := len(gwNodes.Items)
+
+	if numGatewayNodes == input.Gateways {
+		reporter.Succeeded("Current gateways match the required number of gateways")
+		return nil
+	}
+
+	// Currently, we only support increasing the number of Gateway nodes which could be a valid use-case
+	// to convert a non-HA deployment to an HA deployment. We are not supporting decreasing the Gateway
+	// nodes (for now) as it might impact the datapath if we accidentally delete the active GW node.
+	if numGatewayNodes < input.Gateways {
+		gatewayNodesToDeploy := input.Gateways - numGatewayNodes
+
+		workerNodes, err := d.K8sClient.ListNodesWithLabel("node-role.kubernetes.io/worker")
+		if err != nil {
+			return errors.WithMessagef(err, "failed to list k8s nodes  in project %q", d.projectID)
+		}
+
+		nodes := workerNodes.Items
+		for i := range nodes {
+			alreadyTagged := nodes[i].GetLabels()[submarinerGatewayNodeTag]
+			if alreadyTagged == "true" {
+				continue
+			}
+
+			reporter.Started(fmt.Sprintf("Configuring worker node %q as gateway node", nodes[i].Name))
+
+			err := d.K8sClient.AddGWLabelOnNode(nodes[i].Name)
+			if err != nil {
+				return errors.WithMessagef(err, "failed to label the node %q as gateway node", nodes[i].Name)
+			}
+
+			if err := d.openGatewayPort(groupName, nodes[i].Name, computeClient); err != nil {
+				return errors.WithMessage(err, "failed to open the gateway port")
+			}
+
+			gatewayNodesToDeploy--
+			if gatewayNodesToDeploy <= 0 {
+				reporter.Succeeded("Successfully deployed gateway node")
+				return nil
+			}
+		}
+
+		if gatewayNodesToDeploy > 0 {
+			reporter.Failed(fmt.Errorf("there are insufficient nodes to deploy the required number of gateways"))
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (d *ocpGatewayDeployer) Cleanup(reporter api.Reporter) error {
+	reporter.Started("Removing the gateway configuration from nodes ")
+
+	computeClient, err := openstack.NewComputeV2(d.Client, gophercloud.EndpointOpts{Region: d.Region})
+	if err != nil {
+		return errors.WithMessagef(err, "error creating the compute client for the region: %q", d.Region)
+	}
+
+	gwNodesList, err := d.K8sClient.ListGatewayNodes()
+	if err != nil {
+		reporter.Failed(err)
```

Comment:
Here we notify the reporter of the error but in other places we don't. We should be consistent either way.